main.nf

```diff
@@ -19,4 +30,39 @@ workflow mtbseq {
 
 }
 
+workflow per_sample {
+    reads_ch = Channel.fromFilePairs(params.reads)
+
+    TBBWA(reads_ch, params.gatk38_jar, params.user)
+    TBREFINE(TBBWA.out.next_step, params.gatk38_jar, params.user)
+    TBPILE(TBREFINE.out.next_step, params.gatk38_jar, params.user)
+    TBLIST(TBPILE.out.next_step, params.gatk38_jar, params.user)
+    TBVARIANTS(TBLIST.out.next_step, params.gatk38_jar, params.user)
+    TBSTATS(TBBWA.out.next_step,TBLIST.out.next_step, params.gatk38_jar, params.user)
+    TBSTRAINS(TBLIST.out.next_step, params.gatk38_jar, params.user)
+
+
+
+}
 
+workflow cohort {
+    reads_ch = Channel.fromFilePairs(params.reads)
+
+    TBBWA(reads_ch, params.gatk38_jar, params.user)
+    TBREFINE(TBBWA.out.next_step, params.gatk38_jar, params.user)
+    TBPILE(TBREFINE.out.next_step, params.gatk38_jar, params.user)
+    TBLIST(TBPILE.out.next_step, params.gatk38_jar, params.user)
+    TBVARIANTS(TBLIST.out.next_step, params.gatk38_jar, params.user)
+    TBSTATS(TBBWA.out.next_step,TBLIST.out.next_step, params.gatk38_jar, params.user)
+    TBSTRAINS(TBLIST.out.next_step, params.gatk38_jar, params.user)
+
+    samples_tsv_file = TBBWA.out.genomes_names
+            .collect()
+            .flatten().map { n -> "$n" + "\t" + "${params.library_name}" + "\n" }
+            .collectFile(name: 'samples.tsv', newLine: false, storeDir: "${params.outdir}")
+
+    TBJOIN(samples_tsv_file,TBVARIANTS.out[0].collect(), TBLIST.out[0].collect(), params.gatk38_jar, params.user)
+    TBAMEND(TBJOIN.out.next_step, params.gatk38_jar, params.user)
+    TBGROUPS(TBAMEND.out.next_step, params.gatk38_jar, params.user)
```

Comment:
@Mxrcon , there is a crucial thing that was missed here. NF does not work serially, it works in parallel which means that as soon STEP_1 could either pass `sample_1` OR `sample_2` to STEP_2.

In such cases, the `join` operator is used to make sure that the sample sample stays in sync across various processes.

Please have a look here https://github.com/ctmrbio/BACTpipe/blob/bfab3877965f1d9c4048d363ba0dcfb4d36f8aaa/bactpipe.nf#L87

Reply:
Are you saying specifically about  `tbjoin`? I think that i didn't understand how to implement that on `TBJOIN()` ,

Reply:
I think that i understood the problem. so I'll resolve, thank you for pointing me this, i haven't thought about it

Reply:
You're welcome :)